### PR TITLE
Add handlers for read, update, and delete to CREATE_YOUR_FIRST_EXTENSION doc

### PR DESCRIPTION
**Pull request by dbwiddis**
### Description

Updates `CREATE_YOUR_FIRST_EXTENSION` with the RUD part of CRUD (and a better C example).  Also fixes the internal links broken in earlier reviews.

Please merge #827 (and then rebase this PR) prior to merging this PR, as this builds on updates from that PR to Named Routes.

Reference project implementing this, with the code tested: https://github.com/dbwiddis/CRUDExtension

### Issues Resolved

Part of #836 (Documentation)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/OpenSearch/blob/main/CONTRIBUTING.md#developer-certificate-of-origin).

### CODE REVIEW

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 1]
```diff
@@ -147,53 +147,74 @@ public class CRUDExtension extends BaseExtension implements ActionExtension {
...
+    Function<RestRequest, RestResponse> deleteHandler = (request) -> {
         return new ExtensionRestResponse(request, RestStatus.OK, "To be implemented");
     };
 }
 ```
 
 ## Use OpenSearch clients to implement functionality
 
-During the initial creation of the extension, you either implemented `setExtensionsRunner()` yourself or used the `BaseExtension` class, which does it for you. This gives you access to the `ExtensionsRunner` object that is running this extension. The `ExtensionsRunner` has getters that provide access to many objects you will need, one of which is the `SDKClient`. The `SDKClient` class allows initialization of the OpenSearch Java client, which has synchronous and asynchronous versions. For simplicity, this example uses the synchronous client.
+To use the OpenSearch REST API, you will need an instance of the OpenSearch Java Client.
```
Comment (owaiskazi19) [inline CREATE_YOUR_FIRST_EXTENSION.md:213]: SDKClient as a wrapper of Java Client?

**Review by owaiskazi19 — DISMISSED** [review]

File: CREATE_YOUR_FIRST_EXTENSION.md (hunk 1 above)
Comment (dbwiddis) [reply CREATE_YOUR_FIRST_EXTENSION.md:213]: Not sure what the question is?  

SDKClient doesn't wrap Java Client, it just gives a one-liner initialization option to a non-secure client, useful for tutorials like this.

Comment (owaiskazi19) [reply CREATE_YOUR_FIRST_EXTENSION.md:213]: Should we mention? 
```suggestion
To use the OpenSearch REST API, you will need an instance of the SDKClient
```

Comment (dbwiddis) [reply CREATE_YOUR_FIRST_EXTENSION.md:213]: No, we do not need the SDKClient.  

Users can create their own OpenSearchClient. I linked to two options that are in the OpenSearch Client documentation and that is the preferred way to do it, particularly if the users need TLS/SSL.

Comment (owaiskazi19) [reply CREATE_YOUR_FIRST_EXTENSION.md:213]: On line 217, it's mentioned 
``SDKClient` class allows initialization of the OpenSearch Java client`

Comment (dbwiddis) [reply CREATE_YOUR_FIRST_EXTENSION.md:213]: > Can we add java client and SDKClient as 2 options then?

I thought that's exactly what I'm doing here.  I cite the Java Client docs and then I explain how to do it (currently) with SDKClient.

> # Use OpenSearch clients to implement functionality
> 
> To use the OpenSearch REST API, you will need an instance of the OpenSearch Java Client.
> 
> If you require SSL and TLS capability, refer to the OpenSearch Java Client documentation for either [Apache HttpClient 5 Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-apache-httpclient-5-transport) or [OpenSearch RestClient Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-restclient-transport).
> 
> The `SDKClient` class allows initialization of the OpenSearch Java client ...

Now I may be wrong or imprecise about SSL/TLS.  Whatever we're doing is not secure.

Personally I'd like to not mention SDKClient in this doc at all.  Point them to the official docs for the Java client and don't try to include that in this sample.

Comment (dbwiddis) [reply CREATE_YOUR_FIRST_EXTENSION.md:214]: I've updated this to not even mention SDKClient or how to get a copy of it.  We essentially implement a non-secure version of what's on the docs and we shouldn't point people to that non-production code.  Further, I'm arguing elsewhere (#612, #625) that we should remove it.

That section now reads:

> ## Use OpenSearch Java Client to implement functionality
> 
> To use the OpenSearch REST API, you will need an instance of the OpenSearch Java Client.
> 
> Refer to the OpenSearch Java Client documentation for either [Apache HttpClient 5 Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-apache-httpclient-5-transport) or [OpenSearch RestClient Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-restclient-transport).
> 
> The remainder of this example assumes you have implemented one of the above options in your constructor:
> 
> ```java
> private final OpenSearchClient client;
> 
> public CrudAction() {
>     final OpenSearchTransport transport = // implement per documentation
>     this.client = new OpenSearchClient(transport);
> }
> ```

**Review by owaiskazi19 — APPROVED** [review]

**Review by joshpalis — APPROVED** [review]

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 2]
```diff
@@ -2,13 +2,14 @@
 
 * [Initial setup](#initial-setup)
 * [Obtain network address and port information](#obtain-network-address-and-port-information)
-* [Implement the _`Extension`_ interface](#implement-the-extension-interface)
+* [Implement the _`Extension`_ interface](#implement-the-_-extension-_-interface)
```
Comment (owaiskazi19) [inline CREATE_YOUR_FIRST_EXTENSION.md:5]: The link to this section is broken.

Comment (kolchfa-aws) [reply CREATE_YOUR_FIRST_EXTENSION.md:5]:
```suggestion
* [Implement the _`Extension`_ interface](#implement-the-extension-interface)
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 3]
```diff
@@ -191,32 +209,44 @@ public class CrudAction extends BaseExtensionRestHandler {
...
 
-During the initial creation of the extension, you either implemented `setExtensionsRunner()` yourself or used the `BaseExtension` class, which does it for you. This gives you access to the `ExtensionsRunner` object that is running this extension. The `ExtensionsRunner` has getters that provide access to many objects you will need, one of which is the `SDKClient`. The `SDKClient` class allows initialization of the OpenSearch Java client, which has synchronous and asynchronous versions. For simplicity, this example uses the synchronous client.
+Refer to the OpenSearch Java Client documentation for either [Apache HttpClient 5 Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-apache-httpclient-5-transport) or [OpenSearch RestClient Transport](https://opensearch.org/docs/latest/clients/java/#initializing-the-client-with-ssl-and-tls-enabled-using-restclient-transport).
 
-First, update `CRUDExtension` to send a copy of this `ExtensionsRunner` instance to our handler class:
+The remainder of this example assumes you have implemented one of the above options in your constructor:
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:218]:
```suggestion
The remainder of this example assumes you have implemented one of the preceding options in your constructor:
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 4]
```diff
@@ -191,32 +209,44 @@ public class CrudAction extends BaseExtensionRestHandler {
...
+public CrudAction() {
+    final OpenSearchTransport transport = // implement per documentation
+    this.client = new OpenSearchClient(transport);
 }
 ```
 
-Next, you'll create an instance field for the client in the handler class and set its value in the constructor:
+### Defining a Document class
+
+For our CRUD sample we will create a simple Java class with a single field.
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:231]:
```suggestion
For your CRUD sample, you will create a simple Java class with a single field:
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 5]
```diff
@@ -191,32 +209,44 @@ public class CrudAction extends BaseExtensionRestHandler {
...
+    public void setValue(String value) {
+        this.value = value;
+    }
 }
 ```
 
 ### Creating (PUT) a document in an index
 
-Now in our handler function, create an index (if it doesn't exist):
+Now in our create handler function, create an index (if it doesn't exist):
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:249]:
```suggestion
Now in the create handler function, create an index (if it doesn't exist):
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 6]
```diff
@@ -245,34 +275,103 @@ The user needs the ID of the created document (`response.id()`) for further hand
...
+        return createJsonResponse(request, RestStatus.OK, "_id", response.id());
+    }
+    return createJsonResponse(request, RestStatus.INTERNAL_SERVER_ERROR, "failed", response.result().toString());
 };
 ```
 
 ### Reading (GET) a document in an index
 
-TBD
+We can now use the read handler function to get the document we just created, using its ID, which we will pass as a named parameter in the path.  We can then get the document by ID.
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:305]:
```suggestion
You can now use the read handler function to get the document we just created, using its ID, which you will pass as a named parameter in the path.  You can then get the document by ID.
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 7]
```diff
@@ -245,34 +275,103 @@ The user needs the ID of the created document (`response.id()`) for further hand
...
+
+```java
+String id = request.param("id");
+GetResponse<CrudData> response = client.get(new GetRequest.Builder().index("crudsample").id(id).build(), CrudData.class);
+```
+
+Adding in exception handling, the full handler method is:
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:312]:
```suggestion
Adding exception handling, the following is the full handler method:
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 8]
```diff
@@ -245,34 +275,103 @@ The user needs the ID of the created document (`response.id()`) for further hand
...
+        return createJsonResponse(request, RestStatus.OK, "value", response.source().getValue());
+    }
+    return createJsonResponse(request, RestStatus.NOT_FOUND, "error", "not_found");
+};
+```
 
 ### Updating (POST) a document in an index
 
-TBD
+We will create a new document similar to what we did in the create handler, and parse the ID as we did in the read handler, and then update that document.  With exception handling, the update handler method is:
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:333]:
```suggestion
You will create a new document similar to what you did in the create handler, parse the ID as you did in the read handler, and then update that document.  With exception handling, the following is the update handler method:
```

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 9]
```diff
@@ -245,34 +275,103 @@ The user needs the ID of the created document (`response.id()`) for further hand
...
+        return createEmptyJsonResponse(request, RestStatus.OK);
+    }
+    return createJsonResponse(request, RestStatus.INTERNAL_SERVER_ERROR, "failed", response.result().toString());
+};
+```
 
 ### Deleting (DELETE) a document in an index
 
-TBD
+We only need the ID to delete a document, so the delete handler method is:
```
Comment (kolchfa-aws) [inline CREATE_YOUR_FIRST_EXTENSION.md:360]:
```suggestion
You only need the ID to delete a document, so the delete handler method is implemented as follows:
```

**Review by kolchfa-aws — COMMENTED** [review]
A couple of comments, mainly making the document more user-centric. Thanks!

**Comment by dbwiddis** [comment]
> A couple of comments, mainly making the document more user-centric. Thanks!

Thanks!  Can you tell I spent way too long in academia where "we" is the standard? :)

File: CREATE_YOUR_FIRST_EXTENSION.md [hunk 10]
```diff
@@ -147,17 +148,18 @@ public class CRUDExtension extends BaseExtension implements ActionExtension {
 
 These classes must implement _`ExtensionRestHandler`_, which is a functional interface that requires the implementation of the `handleRequest()` method with the signature `public ExtensionRestResponse handleRequest(RestRequest request)`.
 
-The `BaseExtensionRestHandler` class provides many useful methods for exception handling in requests, including a `RouteHandler` class that eases logical separation of multiple `Route` choices.
+The `BaseExtensionRestHandler` class provides many useful methods for exception handling in requests.
 
-For the CRUD extension example, you'll implement one REST route for each option and delegate it to the appropriate handler function, although each one could be stored in its own file:
+For the CRUD extension example, you'll implement one REST route for each option and delegate it to the appropriate handler function.  Each route is an instance of `NamedRoute` and requires at least a method, path, and globally unique name.
```
[the remark this answers is not in the document]
Comment (dbwiddis) [reply CREATE_YOUR_FIRST_EXTENSION.md:153]: https://www.theatlantic.com/science/archive/2018/05/two-spaces-after-a-period/559304/

[the remark this answers is not in the document]
Comment (natebower) [reply CREATE_YOUR_FIRST_EXTENSION.md:153]: Please use one space after a period, in accordance with our style. Thanks!

**Comment by kolchfa-aws** [comment]
Yes, "we" when we're doing an experiment, and "you" when you want the user to do something :)

File: CREATE_YOUR_FIRST_EXTENSION.md (hunk 2 above)
Comment (kolchfa-aws) [reply CREATE_YOUR_FIRST_EXTENSION.md]: If you use IDEA, you can generate the TOC automatically :) https://www.jetbrains.com/help/idea/markdown.html#table-of-contents

**Review by owaiskazi19 — APPROVED** [review]

**Comment by owaiskazi19** [comment]
@kolchfa-aws @natebower looks good to you before we merge the changes in?

**Pull request merged**